core/src/main/java/org/elasticsearch/index/query/NotQueryBuilder.java

```diff
@@ -35,24 +39,34 @@
 
     private String queryName;
 
-    static final NotQueryBuilder PROTOTYPE = new NotQueryBuilder();
+    static final NotQueryBuilder PROTOTYPE = new NotQueryBuilder(null);
 
     public NotQueryBuilder(QueryBuilder filter) {
-        this.filter = Objects.requireNonNull(filter);
```

Comment:
I don't think so, unfortunately parseInnerFilter/Builder() can legally return `null` and also does this when the inner query e.g. is `{ "constant_score" : filter {} }`. This is currently legal rest query which parses to `null` in perseInnerXxx(), this is then passed up by enclosing filters/queries until it gets ignored somewhere. 
The former existing null-check was okay for the existing builder-pattern, but now that we use it for intermediate query representation as well, I think this needs to be allowed.